File: problems-adapt/0142_cycle-entry-in-a-list/solution_floyd.ts
interface CycleNode {
    val: number;
    next: CycleNode | null;
}

function listCycleEntry(values: number[], tailLink: number): number {
    if (values.length === 0) {
        return -1;
    }
    // Materialize the wire form: one node per value, then close the cycle.
    const nodes: CycleNode[] = values.map((v) => ({ val: v, next: null }));
    for (let i = 0; i < nodes.length - 1; i++) {
        nodes[i].next = nodes[i + 1];
    }
    if (tailLink !== -1) {
        nodes[nodes.length - 1].next = nodes[tailLink];
    }
    // Phase 1: tortoise-and-hare scan; fast falling off the end means no cycle.
    let slow: CycleNode | null = nodes[0];
    let fast: CycleNode | null = nodes[0];
    while (fast !== null && fast.next !== null) {
        slow = slow!.next;
        fast = fast.next.next;
        if (slow === fast) {
            // Phase 2: with a = head-to-entry, b = entry-to-meeting and
            // c = the rest of the loop, a + 2b + c = 2(a + b) gives c = a,
            // so a finder restarted at the head and slow continuing from
            // the meeting point converge after exactly a steps — on the
            // entry node.
            let finder: CycleNode | null = nodes[0];
            while (finder !== slow) {
                finder = finder!.next;
                slow = slow!.next;
            }
            // The judge wants an index: count steps from head to entry.
            let index = 0;
            let entry: CycleNode | null = nodes[0];
            while (entry !== finder) {
                entry = entry!.next;
                index++;
            }
            return index;
        }
    }
    return -1;
}

File: problems-adapt/0001-0100/0054_cycle-entry-in-a-list/solution_floyd.cpp
class Solution {
  private:
    struct ListNode {
        int val;
        ListNode *next;
        ListNode(int v) : val(v), next(nullptr) {}
    };

  public:
    int listCycleEntry(vector<int> &values, int tailLink) {
        if (values.empty()) {
            return -1;
        }
        // Materialize the wire form: one node per value, then close the cycle.
        vector<ListNode *> nodes;
        nodes.reserve(values.size());
        for (int v : values) {
            nodes.push_back(new ListNode(v));
        }
        for (size_t i = 0; i + 1 < nodes.size(); i++) {
            nodes[i]->next = nodes[i + 1];
        }
        if (tailLink != -1) {
            nodes.back()->next = nodes[tailLink];
        }
        // Phase 1: tortoise-and-hare scan; fast falling off the end means
        // no cycle.
        ListNode *slow = nodes[0];
        ListNode *fast = nodes[0];
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                // Phase 2: with a = head-to-entry, b = entry-to-meeting and
                // c = the rest of the loop, a + 2b + c = 2(a + b) gives c = a,
                // so a finder restarted at the head and slow continuing from
                // the meeting point converge after exactly a steps — on the
                // entry node.
                ListNode *finder = nodes[0];
                while (finder != slow) {
                    finder = finder->next;
                    slow = slow->next;
                }
                // The judge wants an index: count steps from head to entry.
                int index = 0;
                ListNode *entry = nodes[0];
                while (entry != finder) {
                    entry = entry->next;
                    index++;
                }
                for (ListNode *n : nodes) {
                    delete n;
                }
                return index;
            }
        }
        for (ListNode *n : nodes) {
            delete n;
        }
        return -1;
    }
};
